ee/tabby-webserver/src/schema/settings.rs

```diff
@@ -0,0 +1,55 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use juniper::GraphQLObject;
+use validator::{Validate, ValidationError};
+
+#[async_trait]
+pub trait SettingService: Send + Sync {
+    async fn read_server_setting(&self) -> Result<ServerSetting>;
+    async fn update_server_setting(&self, setting: ServerSetting) -> Result<()>;
+}
+
+#[derive(GraphQLObject)]
+pub struct ServerSetting {
+    pub security_allowed_register_domain_list: Vec<String>,
+    pub security_disable_client_side_telemetry: bool,
+    pub network_external_url: String,
+}
+
+#[derive(Validate)]
+pub struct ServerSettingValidation<'a> {
+    #[validate(custom = "validate_urls")]
+    pub security_allowed_register_domain_list: Vec<&'a str>,
+    #[validate(url)]
+    pub network_external_url: &'a str,
+}
+
+fn validate_urls(urls: &Vec<&str>) -> Result<(), ValidationError> {
+    for url in urls {
+        if !validate_domain(*url) {
+            return Err(ValidationError::new("invalid_url"));
+        }
+    }
+    Ok(())
+}
+
+fn validate_domain(s: &str) -> bool {
```

Comment:
I know this file isn't the right place for this, but wasn't sure where to put it, either. My other PR will need this function as well.

Reply:
consider `src/validators.rs`

I also recommend borrow the email validator: https://github.com/Keats/validator/blob/5dec97c3965af831b2715930341f79fb6d02f6fb/validator/src/validation/email.rs#L24

so you delegate the call `validate_email("noreply@" + domain)`